Implementing CounterStyle representation from StyleRuleCounterStyle

vitorroriz opened the pull request:
#### ac2f8a4229ceb0edf14da7a59709f299970836b4
<pre>
Implementing CounterStyle representation from StyleRuleCounterStyle
<a href="https://bugs.webkit.org/show_bug.cgi?id=248666">https://bugs.webkit.org/show_bug.cgi?id=248666</a>
rdar://102910060

Reviewed by Antti Koivisto.
Reviewed by Antti Koivisto.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Adding CSSCounterStyle and CSSCounterStyleDescriptors .h and .cpp files.

* Source/WebCore/css/CSSCounterStyle.cpp: Added.
(WebCore::CounterStyle::isInRange const):
Returns true if counter is in range, as defined by
<a href="https://www.w3.org/TR/css-counter-styles-3/#descdef-counter-style-range.">https://www.w3.org/TR/css-counter-styles-3/#descdef-counter-style-range.</a>

(WebCore::CounterStyle::CounterStyle):
(WebCore::CounterStyle::create):

(WebCore::CounterStyle::createCounterStyleDecimal):
Create a CounterStyle object representing a Decimal counter.

(WebCore::CounterStyle::setFallbackReference):

(WebCore::CounterStyle::extendAndResolve):
Extend another CounterStyle as defined by
<a href="https://www.w3.org/TR/css-counter-styles-3/#extends-system">https://www.w3.org/TR/css-counter-styles-3/#extends-system</a> and resolve
its Extend reference. The counter&apos;s system value is then promoted
from &apos;extend&apos; to the valye of the CounterStyle we are extending.

(WebCore::CounterStyle::text):
Dummy text representation for the counter. The real representation wil
be implemented by rdar://103648354.

* Source/WebCore/css/CSSCounterStyle.h: Added.
(WebCore::CounterStyle::operator== const):
(WebCore::CounterStyle::fallbackName const):
(WebCore::CounterStyle::extendsName const):
(WebCore::CounterStyle::isFallbackUnresolved):
(WebCore::CounterStyle::isExtendsUnresolved):
(WebCore::CounterStyle::isExtendsSystem const):
(WebCore::CounterStyle::system const):
(WebCore::CounterStyle::isPredefinedCounterStyle const):
Returns true if CounterStyle is predefined by the user agent stylesheet.
A counter style parsed from author-defined rule will return false.

(WebCore::CounterStyle::isAutoRange const):
Returns true if the range descriptor is set to &apos;auto&apos;. A &apos;auto&apos; value
is represented by a empty vector of ranges.

* Source/WebCore/css/CSSCounterStyleRule.cpp:
(WebCore::toCounterStyleSystemEnum):
* Source/WebCore/css/CSSCounterStyleRule.h:
Moving enum to CSSCounterStyle.h.
Adding getters for the properties&apos;s CSS values.

* Source/WebCore/css/CSSStyleSheet.cpp:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterStyleSystem):
* Source/WebCore/dom/Element.h:
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
(WebCore::Style::MatchedDeclarationsCache::isCacheable):
Adding FIXME with matching radars.

CSSCounterStyleDescriptors defines the descriptors for the counter-style
rule in a specialized manner. It serves as a payload for CSSCounterStyle.

* Source/WebCore/css/CSSCounterStyleDescriptors.cpp: Added.
(WebCore::translateRangeFromStyleProperties):
(WebCore::symbolToString):
(WebCore::translatePadFromStyleProperties):
(WebCore::translateNegativeSymbolsFromStyleProperties):
(WebCore::translateSymbolsFromStyleProperties):
(WebCore::translateFallbackNameFromStyleProperties):
(WebCore::translatePrefixFromStyleProperties):
(WebCore::translateSuffixFromStyleProperties):
(WebCore::extractDataFromSystemDescriptor):
(WebCore::CSSCounterStyleDescriptors::setExplicitlySetDescriptors):
Translation from CSSValues of rule&apos;s properties to adequated
specialized types.
(WebCore::CSSCounterStyleDescriptors::create):
Creates an instance of CSSCounterStyleDescriptor based on
StyleProperties. A create function was preferred instead of a
constructor in order to maintain the class as a aggregate, enabling
desginated initializers.
* Source/WebCore/css/CSSCounterStyleDescriptors.h: Added.

Canonical link: <a href="https://commits.webkit.org/258434@main">https://commits.webkit.org/258434@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3006368c54c55e4458f383eb89223c673d00771a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101945 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11090 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35017 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111277 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/171478 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105925 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12056 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2006 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94369 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109031 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107725 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9227 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92496 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/37088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91111 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23946 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4674 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25407 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4763 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1848 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10839 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44895 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5788 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6513 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -0,0 +1,397 @@
+{ }
+
+CounterStyle::CounterStyle(const StyleRuleCounterStyle& rule)
+: m_name { rule.name() }, m_system { toCounterStyleSystemEnum(rule.systemCSSValue()) }, m_negativeSymbols { translateNegativeSymbolsFromStyleRule(rule) }, m_prefix { translatePrefixFromStyleRule(rule) }, m_suffix { translateSuffixFromStyleRule(rule) }, m_ranges { translateRangeFromStyleRule(rule) }, m_pad { translatePadFromStyleRule(rule) }, m_fallbackName { translateFallbackNameFromStyleRule(rul ...
```
That's a lot of constructor parameters. Maybe this type could be a struct (with public fields) instead of a class? Then you just initialize it with { ... } without writing an explicit constructor.

vitorroriz replied:
CounterStyle is a derived class so it won't be classified as an aggregate, right?
https://en.cppreference.com/w/cpp/language/aggregate_initialization#Designated_initializers

anttijk replied:
Right but the question is if should/needs to be refcounted. Alternatively you could have a separate pure struct that just contains the fields. CounterStyleDescriptors or something.

anttijk replied:
The formatting does not follow WebKit style. Each field initialization should be in a separate line, indented.

anttijk commented on Source/WebCore/css/CSSCounterStyleRule.h:
```diff
@@ -44,6 +44,15 @@ enum class CounterStyleSystem : uint8_t {
+    CounterStyleNameReference,
+};
+
 class StyleRuleCounterStyle final : public StyleRuleBase {
```
This should really be in a file of its own.

anttijk commented on Source/WebCore/css/CSSCounterStyleRule.h:
```diff
@@ -66,6 +76,17 @@ class StyleRuleCounterStyle final : public StyleRuleBase {
+    RefPtr<CSSValue> fallbackCSSValue() const { return getPropertyCSSValue(CSSPropertyFallback); }
+    RefPtr<CSSValue> symbolsCSSValue() const { return getPropertyCSSValue(CSSPropertySymbols); }
+    RefPtr<CSSValue> additiveSymbolsCSSValue() const { return getPropertyCSSValue(CSSPropertyAdditiveSymbols); }
+    RefPtr<CSSValue> speakAsCSSValue() const { return getPropertyCSSValue(CSSPropertySpeakAs); }
```
I know you didn't implement this type but the factoring here is really not ideal. Instead of a StyleProperties member there should be a more processed type that represents the counter style as enums/OptionSets or whatever is appropriate of each type.

You have such a type. Maybe you can replace StyleProperties with your CounterStyle type here?

anttijk commented on Source/WebCore/css/CSSCounterStyle.h:
```diff
@@ -0,0 +1,177 @@
+    CounterStyleMap m_authorCounterStyles;
+};
+
+class CounterStyle : public RefCounted<CounterStyle>, public CanMakeWeakPtr<CounterStyle> {
```
This should be in a file of its own.

anttijk replied:
Filename should match the class name (CSSCounterStyle vs CounterStyle)

anttijk commented on Source/WebCore/css/CSSCounterStyle.h:
```diff
@@ -0,0 +1,177 @@
+class CounterStyle;
+using CounterStyleMap = HashMap<AtomString, RefPtr<CounterStyle>>;
+
+class CounterStyleManager : public RefCounted<CounterStyleManager> {
```
We don't like Managers. How about Registry?

This should be in a file of its own.

anttijk commented on Source/WebCore/css/CSSCounterStyle.h:
```diff
@@ -0,0 +1,177 @@
+    WeakPtr<const CounterStyle> m_extendsReference { nullptr };
+    int m_fixedSystemFirstSymbolValue { 1 };
+
+    OptionSet<ExplicitlySetDescriptors> m_explicitlySetDescriptors;
```
Instead of maintaining this OptionSet maybe you could make the individual fields std::optional<>?

vitorroriz replied:
It doesn't achieve the same intent. All the descriptors values are defined, implicitly or explicitly. No explicit values should not be overridden when resolving Extends. I.E.: when CounterStyleA extends CounterStyleB, A will just inherit the values from the descriptors that were explicitly defined by the author, but all the descriptors have at least an implicit value.

anttijk commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -0,0 +1,287 @@
+    auto& list = downcast<CSSValueList>(*ranges);
+    CounterStyle::Ranges result;
+    for (auto& rangeValue : list) {
+        ASSERT(rangeValue->isPrimitiveValue());
```
Having to parse these from a style declaration in StyleRuleCounterStyle is awkward. I think we could have done this at parse time and saved a CounterStyle (or a type that just contains the parsed fields, CounterStyleDescriptor or something) in StyleRuleCounterStyle.

It is bit unclear to me what CounterStyle class represents. Is it final computed counter style that can be pointed to by RenderStyle and directly used by layout/rendering code? Is it an intermediate type that acts as input for style resolution? Or both?

anttijk commented on Source/WebCore/css/CSSCounterStyle.h:
```diff
@@ -0,0 +1,174 @@
+        && m_speakAs == other.m_speakAs
+        && m_speakAsNameReference == other.m_speakAsNameReference
+        && m_predefinedCounterStyle == other.m_predefinedCounterStyle
+        && m_explicitlySetDescriptors == other.m_explicitlySetDescriptors;
```
&& lines should be indented.

anttijk commented on Source/WebCore/css/CSSCounterStyle.h:
```diff
@@ -0,0 +1,174 @@
+    struct Pad {
+        unsigned m_padMinimumLength = 0;
+        Symbol m_padSymbol = "-"_s;
+        bool operator==(const Pad& other) const = default;
```
Unfortunately I think these still fail to compile in some supported linux configurations.

anttijk commented on Source/WebCore/css/CSSCounterStyle.h:
```diff
@@ -0,0 +1,174 @@
+    String initialRepresentation(int);
+
+    CounterStyleSystem system() const { return m_system; }
+    String counterForSystemCyclic(int);
+    String counterForSystemFixed(int);
+    String counterForSystemSymbolic(int);
+    String counterForSystemAlphabetic(int);
+    String counterForSystemNumeric(int);
+    String counterForSystemAdditive(int);
```
These are declared but not defined. It would be nicer to at least add stubs.

anttijk commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -0,0 +1,287 @@
+    for (auto& symbolValue : list) {
+        auto symbolString = symbolToString(&symbolValue.get());
+        if (!symbolString.isNull())
+            result.constructAndAppend(WTFMove(symbolString));
```
Why constructAndAppend instead of just append?

anttijk commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -0,0 +1,287 @@
+        if (high->isInteger())
+            convertedHigh = high->intValue();
+        std::pair<int, int> newRange { convertedLow, convertedHigh };
+        result.constructAndAppend(WTFMove(newRange));
```
Why constructAndAppend instead of just append?

anttijk commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -0,0 +1,287 @@
+        m_explicitlySetDescriptors.add(ExplicitlySetDescriptors::AdditiveSymbols);
+    if (rule.speakAsCSSValue())
+        m_explicitlySetDescriptors.add(ExplicitlySetDescriptors::SpeakAs);
+}
```
It might be better to not try to implement all descriptors in one go. Maybe just start with a few of the most common ones and add the rest when the whole mechanism is in place? It would make these patches easier to follow and review.

vitorroriz committed:
Implementing CounterStyle representation from StyleRuleCounterStyle
https://bugs.webkit.org/show_bug.cgi?id=248666
rdar://102910060

Reviewed by Antti Koivisto.
Reviewed by Antti Koivisto.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Adding CSSCounterStyle and CSSCounterStyleDescriptors .h and .cpp files.

* Source/WebCore/css/CSSCounterStyle.cpp: Added.
(WebCore::CounterStyle::isInRange const):
Returns true if counter is in range, as defined by
https://www.w3.org/TR/css-counter-styles-3/#descdef-counter-style-range.

(WebCore::CounterStyle::CounterStyle):
(WebCore::CounterStyle::create):

(WebCore::CounterStyle::createCounterStyleDecimal):
Create a CounterStyle object representing a Decimal counter.

(WebCore::CounterStyle::setFallbackReference):

(WebCore::CounterStyle::extendAndResolve):
Extend another CounterStyle as defined by
https://www.w3.org/TR/css-counter-styles-3/#extends-system and resolve
its Extend reference. The counter's system value is then promoted
from 'extend' to the valye of the CounterStyle we are extending.

(WebCore::CounterStyle::text):
Dummy text representation for the counter. The real representation wil
be implemented by rdar://103648354.

* Source/WebCore/css/CSSCounterStyle.h: Added.
(WebCore::CounterStyle::operator== const):
(WebCore::CounterStyle::fallbackName const):
(WebCore::CounterStyle::extendsName const):
(WebCore::CounterStyle::isFallbackUnresolved):
(WebCore::CounterStyle::isExtendsUnresolved):
(WebCore::CounterStyle::isExtendsSystem const):
(WebCore::CounterStyle::system const):
(WebCore::CounterStyle::isPredefinedCounterStyle const):
Returns true if CounterStyle is predefined by the user agent stylesheet.
A counter style parsed from author-defined rule will return false.

(WebCore::CounterStyle::isAutoRange const):
Returns true if the range descriptor is set to 'auto'. A 'auto' value
is represented by a empty vector of ranges.

* Source/WebCore/css/CSSCounterStyleRule.cpp:
(WebCore::toCounterStyleSystemEnum):
* Source/WebCore/css/CSSCounterStyleRule.h:
Moving enum to CSSCounterStyle.h.
Adding getters for the properties's CSS values.

* Source/WebCore/css/CSSStyleSheet.cpp:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterStyleSystem):
* Source/WebCore/dom/Element.h:
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
(WebCore::Style::MatchedDeclarationsCache::isCacheable):
Adding FIXME with matching radars.

CSSCounterStyleDescriptors defines the descriptors for the counter-style
rule in a specialized manner. It serves as a payload for CSSCounterStyle.

* Source/WebCore/css/CSSCounterStyleDescriptors.cpp: Added.
(WebCore::translateRangeFromStyleProperties):
(WebCore::symbolToString):
(WebCore::translatePadFromStyleProperties):
(WebCore::translateNegativeSymbolsFromStyleProperties):
(WebCore::translateSymbolsFromStyleProperties):
(WebCore::translateFallbackNameFromStyleProperties):
(WebCore::translatePrefixFromStyleProperties):
(WebCore::translateSuffixFromStyleProperties):
(WebCore::extractDataFromSystemDescriptor):
(WebCore::CSSCounterStyleDescriptors::setExplicitlySetDescriptors):
Translation from CSSValues of rule's properties to adequated
specialized types.
(WebCore::CSSCounterStyleDescriptors::create):
Creates an instance of CSSCounterStyleDescriptor based on
StyleProperties. A create function was preferred instead of a
constructor in order to maintain the class as a aggregate, enabling
desginated initializers.
* Source/WebCore/css/CSSCounterStyleDescriptors.h: Added.

Canonical link: https://commits.webkit.org/258434@main
